### PR TITLE
reduction weights parameter

### LINKED ISSUE #8804: Adding a weights parameter to `reduction`
Hello,

I would like to explore adding a new `weights` parameter to the `dask.array.reduction(x, ...)` function ([dask.array.reduction](https://github.com/dask/dask/blob/main/dask/array/reductions.py#L41)).

This would allow an array of weights to be passed to each chunk, via the `chunk()` callable and alongside the chunk's data.

The use case is to be able to carry out weighted reductions that can't be done "simply", where "simply" means without needing to call `reduction()`. For instance, [weighted averages](https://github.com/dask/dask/blob/main/dask/array/routines.py#L2411) are currently implemented without `reduction()`, and you could easily implement a weighted sum along the same lines.

The particular use case I have is to calculate the unbiased variance of an array weighted with reliability weights. The calculations to do this requires the sum of the weights to be available on each chunk [1], and so the `reduction()` technique must be used.

I have written a PR that implements this in a backwards compatible manner, i.e. all of the existing functions that use `reduction()` (`da.sum`, `da.mean`, `da.var`, etc.) are unaffected by the new code. This PR has no tests yet, as I thought I'd wait to see if there was support this first before writing them :)

The one aspect from the PR that I'd like to highlight here is the requirement for the weights that are passed to `reduction()` to be broadcastable to the data `x`:
```python
     weights : array_like, optional
        Weights to be used in the reduction of `x`. Will be
        automatically broadcast to the shape of `x`, and so must have
        a compatible shape. For instance, if `x` has shape ``(3, 4)``
        then acceptable shapes for `weights` are ``(3, 4)``, ``(4,)``,
        ``(3, 1)``, ``(1, 1)``, ``(1)``, and ``()``.
```
I stipulated this because it is always safe, in that the weights for each chunk are guaranteed to be unambiguous. I don't see this a problematic restriction, though, as the function that calls `reduction()` could easily do a bit of pre-processing on the weights before it passed them to `reduction()` (you would want to do this if, for instance, you re-implemented `da.average` and `da.ma.average` using this functionality, as these functions follow numpy in allowing non-broadcastable weights in some situations).  

Many thanks,
David

[1] e.g. https://en.wikipedia.org/wiki/Weighted_arithmetic_mean#Reliability_weights and https://en.wikipedia.org/wiki/Pooled_variance#Sample-based_statistics

### PR DESCRIPTION
**Pull request by davidhassell**
- [x] Closes #8804
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

~Existing tests pass. Waiting for approval before writing tests for the new functionality. Thanks.~
_Edit: new test now added_

### CODE REVIEW

File: dask/array/reductions.py [hunk 1]
```diff
@@ -52,14 +60,15 @@ def reduction(
...
+    weights=None,
 ):
     """General version of reductions
 
     Parameters
     ----------
     x: Array
         Data being reduced along one or more axes
-    chunk: callable(x_chunk, axis, keepdims)
+    chunk: callable(x_chunk, [weights_chunk=None], axis, keepdims)
```
Comment (jsignell) [inline dask/array/reductions.py:71]: It seems kind of odd to put this in the middle of the existing args.

**Review by jsignell — COMMENTED** [review]
This looks like a very reasonable and isolated change. I just have one question about how the new kwarg will fit in with existing args. Maybe the solution is just, it should go at the end.

File: dask/array/reductions.py (hunk 1 above)
Comment (davidhassell) [reply dask/array/reductions.py:71]: Hi @jsignell  - thanks for looking at this.

I understood (?) that `blockwise` assigns values from its`*args` parameter to the opening parameters of its `func`.

Given that, when we have here  `args=(x, inds, weights, inds)`, `blockwise(chunk, inds, *args, ...)` can only assign `x` and `weights` to the first and second parameters of the  `chunk` function. 

Does that sound right?

**Comment by davidhassell** [comment]
Given the positive feedback so far (thank you!), I've taken the time to write a test ...

**Comment by jsignell** [comment]
Thanks @davidhassell! This is in :)

**Pull request merged**